File: php/tests/Protocols/Mpp/Intent/Session/SessionWireTest.php
<?php

declare(strict_types=1);

namespace PayKit\Tests;

use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use PayKit\Protocols\Mpp\Intent\Session\ClosePayload;
use PayKit\Protocols\Mpp\Intent\Session\CommitPayload;
use PayKit\Protocols\Mpp\Intent\Session\MeteringDirective;
use PayKit\Protocols\Mpp\Intent\Session\OpenPayload;
use PayKit\Protocols\Mpp\Intent\Session\SessionAction;
use PayKit\Protocols\Mpp\Intent\Session\SessionMode;
use PayKit\Protocols\Mpp\Intent\Session\SessionPullVoucherStrategy;
use PayKit\Protocols\Mpp\Intent\Session\SessionRequest;
use PayKit\Protocols\Mpp\Intent\Session\SessionSplit;
use PayKit\Protocols\Mpp\Intent\Session\SignedVoucher;
use PayKit\Protocols\Mpp\Intent\Session\TopUpPayload;
use PayKit\Protocols\Mpp\Intent\Session\VoucherData;

final class SessionWireTest extends TestCase
{
    public function testSessionModeWireValues(): void
    {
        self::assertSame('push', SessionMode::Push->value);
        self::assertSame('pull', SessionMode::Pull->value);
        self::assertSame(SessionMode::Push, SessionMode::from('push'));
    }

    public function testPullVoucherStrategyWireValues(): void
    {
        self::assertSame('clientVoucher', SessionPullVoucherStrategy::ClientVoucher->value);
        self::assertSame('operatedVoucher', SessionPullVoucherStrategy::OperatedVoucher->value);
    }

    public function testSessionRequestOmitsEmptySplitsModesAndNoneFields(): void
    {
        $req = new SessionRequest(cap: '1000', currency: 'USDC', operator: 'op', recipient: 'rec');
        $arr = $req->toArray();
        self::assertArrayNotHasKey('splits', $arr);
        self::assertArrayNotHasKey('modes', $arr);
        self::assertArrayNotHasKey('decimals', $arr);
        self::assertArrayNotHasKey('network', $arr);
        self::assertArrayNotHasKey('description', $arr);
        self::assertArrayNotHasKey('externalId', $arr);
        self::assertArrayNotHasKey('minVoucherDelta', $arr);
    }

    public function testSessionRequestDropsPushOnlyModesList(): void
    {
        $req = new SessionRequest(
            cap: '1000',
            currency: 'USDC',
            operator: 'op',
            recipient: 'rec',
            modes: [SessionMode::Push],
        );
        self::assertArrayNotHasKey('modes', $req->toArray());
    }

    public function testSessionRequestEmitsPushAndPullModesAndStrategy(): void
    {
        $req = new SessionRequest(
            cap: '1000',
            currency: 'USDC',
            operator: 'op',
            recipient: 'rec',
            modes: [SessionMode::Push, SessionMode::Pull],
            pullVoucherStrategy: SessionPullVoucherStrategy::ClientVoucher,
        );
        $arr = $req->toArray();
        self::assertSame(['push', 'pull'], $arr['modes']);
        self::assertSame('clientVoucher', $arr['pullVoucherStrategy']);

        $back = SessionRequest::fromArray($arr);
        self::assertSame([SessionMode::Push, SessionMode::Pull], $back->modes);
        self::assertSame(SessionPullVoucherStrategy::ClientVoucher, $back->pullVoucherStrategy);
    }

    public function testSessionRequestRoundTripsSplitsAndMetadata(): void
    {
        $req = new SessionRequest(
            cap: '10000000',
            currency: 'USDC',
            operator: 'op',
            recipient: 'rec',
            decimals: 6,
            network: 'mainnet-beta',
            splits: [new SessionSplit('s1', 100), new SessionSplit('s2', 200)],
            programId: 'prog123',
            description: 'API session',
            externalId: 'ref-1',
            minVoucherDelta: '500',
        );
        $back = SessionRequest::fromArray($req->toArray());
        self::assertCount(2, $back->splits);
        self::assertSame(100, $back->splits[0]->bps);
        self::assertSame('prog123', $back->programId);
        self::assertSame('ref-1', $back->externalId);
        self::assertSame('500', $back->minVoucherDelta);
        self::assertSame('API session', $back->description);
    }

    public function testOpenPayloadPushFields(): void
    {
        $p = OpenPayload::push('chan1', '1000000', 'signer1', 'txsig');
        self::assertSame(SessionMode::Push, $p->mode);
        self::assertSame('chan1', $p->sessionId());
        self::assertSame('1000000', $p->depositAmount());
        self::assertNull($p->tokenAccount);
    }

    public function testOpenPayloadPullFields(): void
    {
        $p = OpenPayload::pull('tokacct', '5000000', 'wallet1', 'signer1', 'approvesig');
        self::assertSame(SessionMode::Pull, $p->mode);
        self::assertSame('tokacct', $p->sessionId());
        self::assertSame('5000000', $p->depositAmount());
        self::assertSame('wallet1', $p->owner);
        self::assertNull($p->channelId);
    }

    public function testOpenPayloadPushRoundTripsAndOmitsPullFields(): void
    {
        $p = OpenPayload::push('chan1', '1000000', 'signer1', 'txsig');
        $arr = $p->toArray();
        self::assertSame('push', $arr['mode']);
        self::assertSame('chan1', $arr['channelId']);
        self::assertArrayNotHasKey('tokenAccount', $arr);
        $back = OpenPayload::fromArray($arr);
        self::assertSame('chan1', $back->channelId);
    }

    public function testOpenPayloadPullRoundTripsAndOmitsChannelId(): void
    {
        $p = OpenPayload::pull('tokacct', '5000000', 'wallet1', 'signer1', 'approvesig');
        $arr = $p->toArray();
        self::assertSame('pull', $arr['mode']);
        self::assertSame('tokacct', $arr['tokenAccount']);
        self::assertArrayNotHasKey('channelId', $arr);
        $back = OpenPayload::fromArray($arr);
        self::assertSame('tokacct', $back->tokenAccount);
        self::assertSame('wallet1', $back->owner);
    }

    public function testSaltSerializesAsStringAndAcceptsLegacyNumber(): void
    {
        $salt = '18446744073709551608'; // u64::MAX - 7
        $p = OpenPayload::paymentChannel(
            SessionMode::Push,
            'chan1',
            '1000000',
            'payer1',
            'payee1',
            'mint1',
            $salt,
            900,
            'signer1',
            'txsig',
        )->withTransaction('open-tx')->withInitTx('init-tx')->withUpdateTx('update-tx');

        $arr = $p->toArray();
        self::assertSame($salt, $arr['salt']);
        self::assertIsString($arr['salt']);
        self::assertSame('open-tx', $arr['transaction']);
        self::assertSame('init-tx', $arr['initMultiDelegateTx']);
        self::assertSame('update-tx', $arr['updateDelegationTx']);

        // Legacy JSON number salt is accepted on decode and normalized.
        $legacy = OpenPayload::fromArray([
            'mode' => 'push',
            'channelId' => 'chan1',
            'deposit' => '1000000',
            'salt' => 42,
            'gracePeriod' => 900,
            'authorizedSigner' => 'signer1',
            'signature' => 'txsig',
        ]);
        self::assertSame('42', $legacy->salt);
    }

    public function testOpenPayloadMissingModeFailsDecode(): void
    {
        $this->expectException(InvalidArgumentException::class);
        OpenPayload::fromArray([
            'channelId' => 'chan1',
            'deposit' => '1000',
            'authorizedSigner' => 's',
            'signature' => 'sig',
        ]);
    }

    public function testSessionActionOpenPushTag(): void
    {
        $action = SessionAction::open(OpenPayload::push('chan123', '5000000', 'signer123', 'sig456'));
        $arr = $action->toArray();
        self::assertSame('open', $arr['action']);
        self::assertSame('push', $arr['mode']);
        $back = SessionAction::fromArray($arr);
        self::assertSame(SessionAction::OPEN, $back->action);
        self::assertNotNull($back->open);
        self::assertSame('chan123', $back->open->channelId);
    }

    public function testSessionActionVoucherTag(): void
    {
        $voucher = new SignedVoucher(new VoucherData('chan1', '500000', 9999999999, 3), 'sig_here');
        $action = SessionAction::voucher($voucher);
        $arr = $action->toArray();
        self::assertSame('voucher', $arr['action']);
        $back = SessionAction::fromArray($arr);
        self::assertNotNull($back->voucher);
        self::assertSame('500000', $back->voucher->data->cumulative);
        self::assertSame(3, $back->voucher->data->nonce);
    }

    public function testSessionActionCommitTag(): void
    {
        $voucher = new SignedVoucher(new VoucherData('chan1', '500000', 9999999999), 'sig_here');
        $action = SessionAction::commit(new CommitPayload('delivery-1', $voucher));
        $arr = $action->toArray();
        self::assertSame('commit', $arr['action']);
        self::assertSame('delivery-1', $arr['deliveryId']);
        $back = SessionAction::fromArray($arr);
        self::assertNotNull($back->commit);
        self::assertSame('delivery-1', $back->commit->deliveryId);
    }

    public function testSessionActionTopUpTagUsesCapitalU(): void
    {
        $action = SessionAction::topUp(new TopUpPayload('chan1', '9000000', 'txsig'));
        $arr = $action->toArray();
        self::assertSame('topUp', $arr['action']);
        $back = SessionAction::fromArray($arr);
        self::assertNotNull($back->topUp);
        self::assertSame('9000000', $back->topUp->newDeposit);
    }

    public function testSessionActionCloseTagWithAndWithoutVoucher(): void
    {
        $noVoucher = SessionAction::close(new ClosePayload('chan1'));
        $arr = $noVoucher->toArray();
        self::assertSame('close', $arr['action']);
        self::assertArrayNotHasKey('voucher', $arr);

        $withVoucher = SessionAction::close(new ClosePayload(
            'chan1',
            new SignedVoucher(new VoucherData('chan1', '700000', 9999999999, 7), 'final_sig'),
        ));
        $back = SessionAction::fromArray($withVoucher->toArray());
        self::assertNotNull($back->close);
        self::assertNotNull($back->close->voucher);
        self::assertSame('700000', $back->close->voucher->data->cumulative);
    }

    public function testVoucherDataEmitsCumulativeAmountAndReadsCumulativeAlias(): void
    {
        $data = new VoucherData('chan1', '500000', 42);
        self::assertSame('500000', $data->toArray()['cumulativeAmount']);
        self::assertArrayNotHasKey('cumulative', $data->toArray());

        // Decode accepts the legacy `cumulative` field name.
        $alias = VoucherData::fromArray(['channelId' => 'chan1', 'cumulative' => '777', 'expiresAt' => 42]);
        self::assertSame('777', $alias->cumulative);

        // Decode tolerates a JSON number for cumulative.
        $numeric = VoucherData::fromArray(['channelId' => 'chan1', 'cumulativeAmount' => 888, 'expiresAt' => 42]);
        self::assertSame('888', $numeric->cumulative);
    }

    public function testMeteringDirectiveRoundTrips(): void
    {
        $directive = new MeteringDirective('d1', 'chan1', '125', 'USDC', 7, 4_102_444_800, 'https://x/commit');
        $arr = $directive->toArray();
        self::assertSame('d1', $arr['deliveryId']);
        self::assertSame('https://x/commit', $arr['commitUrl']);
        $back = MeteringDirective::fromArray($arr);
        self::assertSame(7, $back->sequence);
        self::assertSame('125', $back->amount);
    }

    public function testDefaultSessionExpiryConstant(): void
    {
        self::assertSame(
            4_102_444_800,
            \PayKit\Protocols\Mpp\Server\Session\SessionServer::DEFAULT_SESSION_EXPIRES_AT,
        );
    }
}

File: ruby/test/session_wire_test.rb
# frozen_string_literal: true

require_relative "test_helper"

# Wire-shape tests for the session intent types. Mirrors the serde tests in
# `rust/crates/mpp/src/protocol/intents/session.rs::tests`.
class SessionWireTest < Minitest::Test
  include RubyMppTestHelpers

  S = ::Mpp::Protocol::Intents::Session

  # ── Mode / strategy serde ──

  def test_mode_constants
    assert_equal "push", S::Mode::PUSH
    assert_equal "pull", S::Mode::PULL
  end

  def test_pull_voucher_strategy_constants
    assert_equal "clientVoucher", S::PullVoucherStrategy::CLIENT_VOUCHER
    assert_equal "operatedVoucher", S::PullVoucherStrategy::OPERATED_VOUCHER
  end

  def test_commit_status_constants
    assert_equal "committed", S::CommitStatus::COMMITTED
    assert_equal "replayed", S::CommitStatus::REPLAYED
  end

  def test_default_session_expires_at_is_below_max_safe_integer
    assert_equal 4_102_444_800, S::DEFAULT_SESSION_EXPIRES_AT
    assert_operator S::DEFAULT_SESSION_EXPIRES_AT, :<, (2**53) - 1
  end

  # ── SessionRequest ──

  def test_session_request_omits_empty_splits_and_modes
    request = S::SessionRequest.new(
      cap: "10000000", currency: "USDC", operator: pubkey(1), recipient: pubkey(2),
      decimals: 6, network: "localnet"
    )
    wire = request.to_h

    refute wire.key?("splits")
    refute wire.key?("modes")
    refute wire.key?("pullVoucherStrategy")
    refute wire.key?("minVoucherDelta")
    assert_equal "10000000", wire["cap"]
    assert_equal 6, wire["decimals"]
  end

  def test_session_request_serializes_splits_and_modes_when_present
    request = S::SessionRequest.new(
      cap: "5000", currency: "USDC", operator: pubkey(1), recipient: pubkey(2),
      splits: [S::SessionSplit.new(recipient: pubkey(3), bps: 1000)],
      modes: [S::Mode::PUSH, S::Mode::PULL],
      pull_voucher_strategy: S::PullVoucherStrategy::OPERATED_VOUCHER,
      min_voucher_delta: "1000", program_id: pubkey(4), external_id: "ref-1"
    )
    wire = request.to_h

    assert_equal [{"recipient" => pubkey(3), "bps" => 1000}], wire["splits"]
    assert_equal ["push", "pull"], wire["modes"]
    assert_equal "operatedVoucher", wire["pullVoucherStrategy"]
    assert_equal "1000", wire["minVoucherDelta"]
    assert_equal pubkey(4), wire["programId"]
    assert_equal "ref-1", wire["externalId"]
  end

  def test_session_request_roundtrip
    request = S::SessionRequest.new(
      cap: "5000", currency: "USDC", operator: pubkey(1), recipient: pubkey(2),
      modes: [S::Mode::PULL], pull_voucher_strategy: S::PullVoucherStrategy::CLIENT_VOUCHER
    )
    back = S::SessionRequest.from_h(request.to_h)

    assert_equal request.to_h, back.to_h
  end

  def test_session_request_carries_recent_blockhash_and_description
    request = S::SessionRequest.new(
      cap: "5000", currency: "USDC", operator: pubkey(1), recipient: pubkey(2),
      description: "metered API", recent_blockhash: pubkey(8)
    )
    wire = request.to_h

    assert_equal "metered API", wire["description"]
    assert_equal pubkey(8), wire["recentBlockhash"]
    assert_equal "metered API", S::SessionRequest.from_h(wire).description
  end

  # ── OpenPayload ──

  def test_open_push_builder
    payload = S::OpenPayload.push(channel_id: pubkey(5), deposit: 1_000_000, authorized_signer: pubkey(6), signature: "sig")

    assert_equal "push", payload.mode
    assert_equal pubkey(5), payload.session_id
    assert_equal 1_000_000, payload.deposit_amount
    wire = payload.to_h
    assert_equal "open", wire["action"]
    assert_equal "1000000", wire["deposit"]
    refute wire.key?("salt")
  end

  def test_open_payment_channel_salt_serializes_as_string
    payload = S::OpenPayload.payment_channel(
      channel_id: pubkey(5), deposit: 1000, payer: pubkey(1), payee: pubkey(2),
      mint: pubkey(3), salt: 18_446_744_073_709_551_615, grace_period: 3600,
      authorized_signer: pubkey(6), signature: "sig"
    )
    wire = payload.to_h

    assert_equal "18446744073709551615", wire["salt"]
    assert_kind_of String, wire["salt"]
    assert_equal 3600, wire["gracePeriod"]
  end

  def test_open_salt_deserializes_from_string_or_number
    from_string = S::OpenPayload.from_h(
      "action" => "open", "mode" => "push", "salt" => "42",
      "authorizedSigner" => pubkey(6), "signature" => "sig", "channelId" => pubkey(5), "deposit" => "1"
    )
    from_number = S::OpenPayload.from_h(
      "action" => "open", "mode" => "push", "salt" => 42,
      "authorizedSigner" => pubkey(6), "signature" => "sig", "channelId" => pubkey(5), "deposit" => "1"
    )

    assert_equal 42, from_string.salt
    assert_equal 42, from_number.salt
    assert_equal "42", from_string.to_h["salt"]
  end

  def test_open_pull_builder_session_id_is_token_account
    payload = S::OpenPayload.pull(
      token_account: pubkey(7), approved_amount: 5000, owner: pubkey(8),
      authorized_signer: pubkey(6), signature: "sig"
    )

    assert_equal "pull", payload.mode
    assert_equal pubkey(7), payload.session_id
    assert_equal 5000, payload.deposit_amount
    assert_equal "5000", payload.to_h["approvedAmount"]
  end

  def test_open_push_missing_channel_id_raises
    payload = S::OpenPayload.new(mode: S::Mode::PUSH, authorized_signer: pubkey(6), signature: "sig", deposit: "1")

    assert_raises(ArgumentError) { payload.session_id }
  end

  # ── Vouchers ──

  def test_voucher_data_serializes_cumulative_amount_only
    data = S::VoucherData.new(channel_id: pubkey(9), cumulative: "100", expires_at: 1234)
    wire = data.to_h

    assert_equal "100", wire["cumulativeAmount"]
    refute wire.key?("cumulative")
    assert_equal 1234, wire["expiresAt"]
  end

  def test_voucher_data_reads_cumulative_alias
    via_alias = S::VoucherData.from_h("channelId" => pubkey(9), "cumulative" => "55", "expiresAt" => 10)
    via_canonical = S::VoucherData.from_h("channelId" => pubkey(9), "cumulativeAmount" => "55", "expiresAt" => 10)

    assert_equal "55", via_alias.cumulative
    assert_equal "55", via_canonical.cumulative
  end

  def test_voucher_data_canonical_name_wins_over_alias
    data = S::VoucherData.from_h(
      "channelId" => pubkey(9), "cumulativeAmount" => "100", "cumulative" => "1", "expiresAt" => 10
    )

    assert_equal "100", data.cumulative
  end

  def test_voucher_data_message_bytes_roundtrips_against_program
    data = S::VoucherData.new(channel_id: pubkey(9), cumulative: "42", expires_at: 1234)
    expected = ::Mpp::Program::PaymentChannels.voucher_message_bytes(channel_id: pubkey(9), cumulative_amount: 42, expires_at: 1234)

    assert_equal expected, data.message_bytes
  end

  def test_signed_voucher_roundtrip
    data = S::VoucherData.new(channel_id: pubkey(9), cumulative: "100", expires_at: 1234, nonce: 3)
    voucher = S::SignedVoucher.new(data: data, signature: "abc")
    back = S::SignedVoucher.from_h(voucher.to_h)

    assert_equal "100", back.data.cumulative
    assert_equal 3, back.data.nonce
    assert_equal "abc", back.signature
  end

  # ── Action tags ──

  def test_topup_action_tag_uses_capital_u
    payload = S::TopUpPayload.new(channel_id: pubkey(5), new_deposit: 2000, signature: "sig")

    assert_equal "topUp", payload.to_h["action"]
    assert_equal "2000", payload.to_h["newDeposit"]
  end

  def test_session_action_dispatch_by_tag
    open_action, open_payload = S::SessionAction.from_h(
      "action" => "open", "mode" => "push", "authorizedSigner" => pubkey(6),
      "signature" => "sig", "channelId" => pubkey(5), "deposit" => "1"
    )
    topup_action, = S::SessionAction.from_h(
      "action" => "topUp", "channelId" => pubkey(5), "newDeposit" => "2", "signature" => "sig"
    )

    assert_equal "open", open_action
    assert_instance_of S::OpenPayload, open_payload
    assert_equal "topUp", topup_action
  end

  def test_session_action_unknown_tag_raises
    assert_raises(ArgumentError) { S::SessionAction.from_h("action" => "bogus") }
  end

  def test_commit_payload_roundtrip
    data = S::VoucherData.new(channel_id: pubkey(9), cumulative: "100", expires_at: 10)
    payload = S::CommitPayload.new(delivery_id: "d-1", voucher: S::SignedVoucher.new(data: data, signature: "s"))
    back = S::CommitPayload.from_h(payload.to_h)

    assert_equal "commit", payload.to_h["action"]
    assert_equal "d-1", back.delivery_id
    assert_equal "100", back.voucher.data.cumulative
  end

  def test_close_payload_omits_voucher_when_absent
    payload = S::ClosePayload.new(channel_id: pubkey(5))

    refute payload.to_h.key?("voucher")
    assert_equal "close", payload.to_h["action"]
  end
end
